src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopier.java

```diff
@@ -0,0 +1,229 @@
+  }
+
+  /**
+   * Begins an asynchronous copy operation between storage accounts. The backend storage service
+   * performs the copy operation. The method returns an instance of the {@link
+   * BlobContainerCopySyncPoller} which facilitates the management and getting the status of the
+   * copy operation.
+   *
+   * @return an instance of {@link BlobContainerCopySyncPoller} that is associated with the copy
+   *     operation.
+   */
+  public BlobContainerCopySyncPoller beginCopyOperation() {
+
+    if (this.sourceClientFactory != null) {
+      logger.info("Starting copy operation using a container as a source");
+      return beginCopyOperationUsingSourceBlobContainerClient();
+    }
+
+    if (StringUtils.isNotBlank(this.sourceBlobUrl)) {
+      logger.info("Starting copy operation using a signed blob URL as the source");
+      return beginCopyOperationUsingBlobSignedUrl();
+    }
+
+    throw new IllegalArgumentException(
+        "The copy operation can't be started. " + "The source information is missing or invalid");
+  }
+
+  public void setDestinationBlobName(String destinationBlobName) {
+    this.destinationBlobName = destinationBlobName;
+  }
+
+  public void setSourceDestinationPairs(
+      List<BlobCopySourceDestinationPair> sourceDestinationPairs) {
+    this.sourceDestinationPairs = sourceDestinationPairs;
+  }
+
+  public void setSourceBlobUrl(String sourceBlobUrl) {
+    this.sourceBlobUrl = sourceBlobUrl;
+  }
+
+  public void setBlobSourcePrefix(String blobSourcePrefix) {
+    this.blobSourcePrefix =
+        Objects.requireNonNull(blobSourcePrefix, "Blob source prefix must not be null");
+  }
+
+  public void setListOperationTimeout(Duration listOperationTimeout) {
+    this.listOperationTimeout = listOperationTimeout;
+  }
+
+  public void setPollingInterval(Duration pollingInterval) {
+    this.pollingInterval = pollingInterval;
+  }
+
+  public void setSourceClientFactory(BlobContainerClientFactory sourceClientFactory) {
+    this.sourceClientFactory = sourceClientFactory;
+  }
+
+  public BlobContainerClientFactory getDestinationClientFactory() {
+    return destinationClientFactory;
+  }
+
+  public String getBlobSourcePrefix() {
+    return blobSourcePrefix;
+  }
+
+  public BlobContainerClientFactory getSourceClientFactory() {
+    return sourceClientFactory;
+  }
+
+  public String getSourceBlobUrl() {
+    return sourceBlobUrl;
+  }
+
+  public String getDestinationBlobName() {
+    return destinationBlobName;
+  }
+
+  public List<BlobCopySourceDestinationPair> getSourceDestinationPairs() {
+    return sourceDestinationPairs;
+  }
+
+  private BlobContainerCopySyncPoller beginCopyOperationUsingBlobSignedUrl() {
+    BlobUrlParts blobUrlParts = BlobUrlParts.parse(this.sourceBlobUrl);
+
+    String sourceBlobName = blobUrlParts.getBlobName();
+    String destinationBlobName = this.destinationBlobName;
+    if (StringUtils.isBlank(this.destinationBlobName)) {
```

Comment:
This is the same pattern that the SDK follows in their builders. The idea is that you want to validate them upon getting them set for optional properties, which indicates that the caller wants to use them, and therefore the caller must provide valid values. If you wait until the build time, it is hard to determine if a null/empty value exists because the caller didn't intend to use the property or a bug/unexpected behavior incorrectly set such the property.